File: services/learning_service.py
from typing import Dict, List, Any
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from collections import Counter, defaultdict
import json
import re

class LearningService:
    def __init__(self):
        # Initialize NLTK resources
        try:
            nltk.data.find('tokenizers/punkt')
            nltk.data.find('corpora/stopwords')
        except LookupError:
            nltk.download('punkt')
            nltk.download('stopwords')
        
        self.stop_words = set(stopwords.words('english'))
        
        # User-specific data storage
        self.user_data: Dict[int, Dict[str, Any]] = {}
        
        # Initialize default user data structure
        self.default_user_data = {
            "topics": [],           # List of topics the user frequently discusses
            "patterns": Counter(),  # Common word patterns in user messages
            "preferences": {        # User preferences learned from interactions
                "formality": 0.5,   # 0 = casual, 1 = formal
                "verbosity": 0.5,   # 0 = concise, 1 = detailed
                "technicality": 0.5 # 0 = simple, 1 = technical
            },
            "message_count": 0
        }

    def _initialize_user(self, user_id: int):
        """Initialize data structure for a new user."""
        if user_id not in self.user_data:
            self.user_data[user_id] = json.loads(json.dumps(self.default_user_data))

    def _extract_topics(self, text: str) -> List[str]:
        """Extract potential topics from text."""
        # Tokenize and remove stopwords
        tokens = word_tokenize(text.lower())
        tokens = [token for token in tokens if token not in self.stop_words 
                 and token.isalnum() and len(token) > 2]
        
        # Count word frequencies
        word_freq = Counter(tokens)
        
        # Return most common words as topics
        return [word for word, count in word_freq.most_common(3) if count > 1]

    def _analyze_patterns(self, text: str) -> Counter:
        """Analyze text for common patterns."""
        # Extract word pairs (bigrams)
        tokens = word_tokenize(text.lower())
        pairs = [f"{tokens[i]} {tokens[i+1]}" for i in range(len(tokens)-1)]
        
        return Counter(pairs)

    def _update_preferences(self, text: str, user_data: Dict[str, Any]):
        """Update user preferences based on message content."""
        # Update formality preference
        formal_indicators = len(re.findall(r'\b(please|kindly|would|could|may|regarding)\b', text.lower()))
        casual_indicators = len(re.findall(r'\b(hey|hi|yeah|cool|awesome|gonna|wanna)\b', text.lower()))
        
        if formal_indicators > casual_indicators:
            user_data["preferences"]["formality"] = min(1.0, user_data["preferences"]["formality"] + 0.05)
        elif casual_indicators > formal_indicators:
            user_data["preferences"]["formality"] = max(0.0, user_data["preferences"]["formality"] - 0.05)
        
        # Update verbosity preference
        words_per_sentence = len(text.split()) / max(1, len(text.split('.')))
        if words_per_sentence > 15:
            user_data["preferences"]["verbosity"] = min(1.0, user_data["preferences"]["verbosity"] + 0.05)
        elif words_per_sentence < 8:
            user_data["preferences"]["verbosity"] = max(0.0, user_data["preferences"]["verbosity"] - 0.05)
        
        # Update technicality preference
        technical_words = len(re.findall(r'\b(api|function|code|data|system|process|technical)\b', text.lower()))
        if technical_words > 0:
            user_data["preferences"]["technicality"] = min(1.0, user_data["preferences"]["technicality"] + 0.05)

    def process_message(self, user_id: int, message: str) -> Dict[str, Any]:
        """Process a user message and update learning data."""
        self._initialize_user(user_id)
        user_data = self.user_data[user_id]
        
        # Extract and update topics
        new_topics = self._extract_topics(message)
        user_data["topics"].extend(new_topics)
        user_data["topics"] = list(set(user_data["topics"]))[-10:]  # Keep last 10 unique topics
        
        # Update pattern frequencies
        user_data["patterns"].update(self._analyze_patterns(message))
        
        # Update preferences
        self._update_preferences(message, user_data)
        
        # Increment message count
        user_data["message_count"] += 1
        
        return self.get_response_context(user_id)

    def get_response_context(self, user_id: int) -> Dict[str, Any]:
        """Get the learning context for generating a response."""
        self._initialize_user(user_id)
        user_data = self.user_data[user_id]
        
        return {
            "topics": user_data["topics"],
            "patterns": dict(user_data["patterns"].most_common(5)),
            "preferences": user_data["preferences"],
            "message_count": user_data["message_count"]
        }

    def adapt_response(self, user_id: int, response: str) -> str:
        """Adapt a response based on learned user preferences."""
        self._initialize_user(user_id)
        prefs = self.user_data[user_id]["preferences"]
        
        # Adjust response based on formality preference
        if prefs["formality"] > 0.7:
            response = response.replace("yeah", "yes")
            response = response.replace("nope", "no")
            response = response.replace("gonna", "going to")
            response = response.replace("wanna", "want to")
        elif prefs["formality"] < 0.3:
            response = response.replace("certainly", "sure")
            response = response.replace("additionally", "also")
            response = response.replace("however", "but")
        
        # Adjust response based on verbosity preference
        if prefs["verbosity"] < 0.3 and len(response) > 100:
            # Attempt to make response more concise
            sentences = response.split('. ')
            if len(sentences) > 1:
                response = '. '.join(sentences[::2]) + '.'
        
        # Adjust response based on technicality preference
        if prefs["technicality"] > 0.7:
            # Add more technical details or terminology
            response = response.replace("use", "utilize")
            response = response.replace("make", "implement")
            response = response.replace("fix", "resolve")
        elif prefs["technicality"] < 0.3:
            # Simplify technical terms
            response = response.replace("utilize", "use")
            response = response.replace("implement", "make")
            response = response.replace("resolve", "fix")
        
        return response

    def clear_user_data(self, user_id: int):
        """Clear all learned data for a user."""
        if user_id in self.user_data:
            self.user_data[user_id] = json.loads(json.dumps(self.default_user_data))

File: src/services/learningService.js
import { LEARNING_SETTINGS } from '../config/config';

// Common English stop words
const STOP_WORDS = new Set([
  'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for', 'from', 'has', 'he',
  'in', 'is', 'it', 'its', 'of', 'on', 'that', 'the', 'to', 'was', 'were',
  'will', 'with', 'the', 'this', 'but', 'they', 'have', 'had', 'what', 'when',
  'where', 'who', 'which', 'why', 'how'
]);

export class LearningService {
  constructor() {
    this.userData = new Map();
    
    this.defaultUserData = {
      topics: [],
      patterns: new Map(),
      preferences: {
        formality: 0.5,
        verbosity: 0.5,
        technicality: 0.5
      },
      messageCount: 0
    };
  }

  initializeUser(userId) {
    if (!this.userData.has(userId)) {
      this.userData.set(userId, JSON.parse(JSON.stringify({
        ...this.defaultUserData,
        patterns: {}
      })));
    }
  }

  tokenize(text) {
    // Simple word tokenization
    return text.toLowerCase()
      .replace(/[^\w\s]/g, '') // Remove punctuation
      .split(/\s+/)            // Split on whitespace
      .filter(word => word.length > 0);
  }

  extractTopics(text) {
    const tokens = this.tokenize(text)
      .filter(token => !STOP_WORDS.has(token) && 
              token.length > 2 && 
              /^[a-z]+$/.test(token));
    
    const wordFreq = new Map();
    tokens.forEach(token => {
      wordFreq.set(token, (wordFreq.get(token) || 0) + 1);
    });

    return Array.from(wordFreq.entries())
      .filter(([_, count]) => count > 1)
      .sort(([_, a], [__, b]) => b - a)
      .slice(0, 3)
      .map(([word]) => word);
  }

  analyzePatterns(text) {
    const tokens = this.tokenize(text);
    const patterns = new Map();

    for (let i = 0; i < tokens.length - 1; i++) {
      const pair = `${tokens[i]} ${tokens[i + 1]}`;
      patterns.set(pair, (patterns.get(pair) || 0) + 1);
    }

    return patterns;
  }

  updatePreferences(text, userData) {
    // Update formality preference
    const formalCount = LEARNING_SETTINGS.formalIndicators
      .filter(word => text.toLowerCase().includes(word)).length;
    const casualCount = LEARNING_SETTINGS.casualIndicators
      .filter(word => text.toLowerCase().includes(word)).length;

    if (formalCount > casualCount) {
      userData.preferences.formality = Math.min(
        1.0, 
        userData.preferences.formality + LEARNING_SETTINGS.preferenceAdjustmentRate
      );
    } else if (casualCount > formalCount) {
      userData.preferences.formality = Math.max(
        0.0, 
        userData.preferences.formality - LEARNING_SETTINGS.preferenceAdjustmentRate
      );
    }

    // Update verbosity preference
    const wordsPerSentence = text.split(/[.!?]+/).map(s => 
      s.trim().split(/\s+/).length
    ).reduce((a, b) => a + b, 0) / Math.max(1, text.split(/[.!?]+/).length);

    if (wordsPerSentence > 15) {
      userData.preferences.verbosity = Math.min(
        1.0, 
        userData.preferences.verbosity + LEARNING_SETTINGS.preferenceAdjustmentRate
      );
    } else if (wordsPerSentence < 8) {
      userData.preferences.verbosity = Math.max(
        0.0, 
        userData.preferences.verbosity - LEARNING_SETTINGS.preferenceAdjustmentRate
      );
    }

    // Update technicality preference
    const technicalWords = LEARNING_SETTINGS.technicalKeywords
      .filter(word => text.toLowerCase().includes(word)).length;

    if (technicalWords > 0) {
      userData.preferences.technicality = Math.min(
        1.0, 
        userData.preferences.technicality + LEARNING_SETTINGS.preferenceAdjustmentRate
      );
    }
  }

  processMessage(userId, message) {
    this.initializeUser(userId);
    const userData = this.userData.get(userId);

    // Extract and update topics
    const newTopics = this.extractTopics(message);
    userData.topics = [...new Set([...userData.topics, ...newTopics])]
      .slice(-LEARNING_SETTINGS.maxTopics);

    // Update pattern frequencies
    const newPatterns = this.analyzePatterns(message);
    newPatterns.forEach((count, pattern) => {
      userData.patterns[pattern] = (userData.patterns[pattern] || 0) + count;
    });

    // Update preferences
    this.updatePreferences(message, userData);

    // Increment message count
    userData.messageCount++;

    return this.getResponseContext(userId);
  }

  getResponseContext(userId) {
    this.initializeUser(userId);
    const userData = this.userData.get(userId);

    // Get top patterns
    const sortedPatterns = Object.entries(userData.patterns)
      .sort(([_, a], [__, b]) => b - a)
      .slice(0, LEARNING_SETTINGS.maxPatterns);

    return {
      topics: userData.topics,
      patterns: Object.fromEntries(sortedPatterns),
      preferences: userData.preferences,
      messageCount: userData.messageCount
    };
  }

  adaptResponse(userId, response) {
    this.initializeUser(userId);
    const prefs = this.userData.get(userId).preferences;

    // Adjust based on formality preference
    if (prefs.formality > 0.7) {
      response = response
        .replace(/yeah/g, "yes")
        .replace(/nope/g, "no")
        .replace(/gonna/g, "going to")
        .replace(/wanna/g, "want to");
    } else if (prefs.formality < 0.3) {
      response = response
        .replace(/certainly/g, "sure")
        .replace(/additionally/g, "also")
        .replace(/however/g, "but");
    }

    // Adjust based on verbosity preference
    if (prefs.verbosity < 0.3 && response.length > 100) {
      const sentences = response.split(". ");
      if (sentences.length > 1) {
        response = sentences.filter((_, i) => i % 2 === 0).join(". ") + ".";
      }
    }

    // Adjust based on technicality preference
    if (prefs.technicality > 0.7) {
      response = response
        .replace(/use/g, "utilize")
        .replace(/make/g, "implement")
        .replace(/fix/g, "resolve");
    } else if (prefs.technicality < 0.3) {
      response = response
        .replace(/utilize/g, "use")
        .replace(/implement/g, "make")
        .replace(/resolve/g, "fix");
    }

    return response;
  }

  clearUserData(userId) {
    this.userData.set(userId, JSON.parse(JSON.stringify({
      ...this.defaultUserData,
      patterns: {}
    })));
  }
}
